### ISSUE TITLE
VideoLearning Project Review

### ISSUE THREAD

**Issue by Noath2302** [ddobric/neocortexapi#25]
Dear Mr.Dobric,
Please set me a call this or next week.
I am working on [project Video Learning](https://github.com/ddobric/neocortexapi/tree/SequenceLearning_ToanTruong/Project12_HTMCLAVideoLearning/HTMVideoLearning) under branch [SequenceLearning_ToanTruong](https://github.com/ddobric/neocortexapi/tree/SequenceLearning_ToanTruong)  

**Overview:**
The Project used Temporal memory to learn the frame of each video sequentially. Then try to predict the video/ frame sequence when a frame is entered by the user. The output can be seen as a video.

Each frame during the training has its own unique framekey in format **_label_videoName_index_** e.g. circle_vd1_1 

There are 2 test: Run1 and Run2
Same:
The training set are labeled folder which contains videos.
They both start with reading the video and learning with HomeostaticPlasticityController until Stable state is reached
Test picture (from Paint) used after the learning
![rth](https://user-images.githubusercontent.com/29188162/132472457-b922c40a-069d-4d4b-95cc-b38aeb5f692a.png)

Diff:
[**Run1:**](https://github.com/ddobric/neocortexapi/blob/560e924e87036197ff881baff69fdcac4e0ce494/Project12_HTMCLAVideoLearning/HTMVideoLearning/HTMVideoLearning/VideoLearning.cs#L18)
Learning with SP+TM
HTMClassifier learns a framekey corresponding to the Active Cells computed from InputBitArray of a frame.

The test ends after 40 consecutive same accuracy read >90% for the whole training set or reach max Cycles
best accuracy 90%, varies from 54% to 90% sometimes it stucks at one accuracy and must be run again or wait for max Cycle.

The user can then drag a picture into the Console to start testing after learning

Predict each frame with its own framekey, then use the predicted framekey as input to predict the next framekey untill there are no more Predicted Cells in computed output.

**current result:**
The Video are shorter than the original because at some frame, there are no Predicted Cells. 
Overlapping also happens, A circle frame that has its circle overlap a triangle in another frame in the training set is predicted as a triangle.

[**Run2:**](https://github.com/ddobric/neocortexapi/blob/560e924e87036197ff881baff69fdcac4e0ce494/Project12_HTMCLAVideoLearning/HTMVideoLearning/HTMVideoLearning/VideoLearning.cs#L368)
Learning with SP+TM
Adapt logic from [SequenceLearning_DUY](https://github.com/perfectaccountname/neocortexapi/blob/master/source/Samples/NeoCortexApiSample/SequenceLearning_DUY.cs)
HTMClassifier learns a key created by previous frames, e.g. 
the key: 
**circle_vd1_27-circle_vd1_28-circle_vd1_29-circle_vd1_0-circle_vd1_1** 
corresponding to the Active Cells computed from InputBitArray of frame **circle_vd1_1**.

The Test ends after reaching Accuracy 90% or more 100 times
best accuracy 93.333%, varies from 77% to 93% sometimes it stucks at one accuracy and must be run again or wait for max Cycle.

The user can then drag a picture into the Console to start testing after learning.

**current result**
The output video has same length as training videos.
The output video loops and is bound to training videos, it can only output frame sequence in one video in training set.

RESULT FILE (The videos cannot be displayed on issue):
[link google drive](https://drive.google.com/drive/folders/1lgrOrUSqFq58gzycoC3DFUjrwdiJwHBg?usp=sharing)

[A Library for converting Video](https://github.com/ddobric/neocortexapi/tree/SequenceLearning_ToanTruong/Project12_HTMCLAVideoLearning/HTMVideoLearning/VideoLibrary) into a sequence of InputBitArrays and vice versa using OpenCv was developed for the Test.

**Comment by Noath2302** [comment]
The submission for the project is on 20 September, I think we can start review and grading now.
you can refer to the [report ](https://github.com/ddobric/neocortexapi/tree/SequenceLearning_ToanTruong/Project12_HTMCLAVideoLearning/HTMVideoLearning). I keep it as a md file.

**Comment by Noath2302** [comment]
I have added The Experiment to Project [NeoCortexApiSample](https://github.com/ddobric/neocortexapi/tree/SequenceLearning_ToanTruong/source/Samples/NeoCortexApiSample).  

The [VideoLearning.cs](https://github.com/ddobric/neocortexapi/blob/SequenceLearning_ToanTruong/source/Samples/NeoCortexApiSample/VideoLearning.cs) comes with 2 additional components: the [HelperFunction.cs](https://github.com/ddobric/neocortexapi/blob/SequenceLearning_ToanTruong/source/Samples/NeoCortexApiSample/HelperFunction.cs) and the [VideoLibrary Project](https://github.com/ddobric/neocortexapi/tree/SequenceLearning_ToanTruong/source/Samples/NeoCortexApiSample/VideoLibrary), which help with printing colored text and reading video with OpenCV2.

To run it you have to modify the Program.cs to change which Experiment you want to use.  There are 2 option in VideoLearning: Run1 and Run2.  
The dataset can be accessed in the [VideoLibrary](https://github.com/ddobric/neocortexapi/tree/SequenceLearning_ToanTruong/source/Samples/NeoCortexApiSample/VideoLibrary), when running the code, you drag the one of the training data folders to the console for its path and press enter. SmallTrainingSet/ is reccomended for running the experiment.
